cli/internal/libvirt/libvirt.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package libvirt
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	docker "github.com/docker/docker/client"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/spf13/afero"
+)
+
+// Runner handles starting and stopping of containerized libvirt instances.
+type Runner struct {
+	nameFile string
+	file     file.Handler
+}
+
+// New creates a new LibvirtRunner.
+func New() *Runner {
+	return &Runner{
+		nameFile: "libvirt.name",
+		file:     file.NewHandler(afero.NewOsFs()),
+	}
+}
+
+// StartLibvirt starts a containerized libvirt instance.
+func (r *Runner) StartLibvirt(ctx context.Context, name, imageName string) error {
+	docker, err := docker.NewClientWithOpts(docker.FromEnv, docker.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+	defer docker.Close()
+
+	containerName := name + "-libvirt"
+	if _, err := docker.ContainerCreate(ctx,
+		&container.Config{
+			Image: imageName,
+		},
+		&container.HostConfig{
+			NetworkMode: container.NetworkMode("host"),
+			Privileged:  true,
+			AutoRemove:  true,
+			Mounts: []mount.Mount{
+				{
+					Type:   mount.TypeBind,
+					Source: "/dev/kvm",
+					Target: "/dev/kvm",
+				},
+			},
+		},
+		nil,
+		nil,
+		containerName,
+	); err != nil {
+		return err
+	}
+	if err := docker.ContainerStart(ctx, containerName, types.ContainerStartOptions{}); err != nil {
+		_ = docker.ContainerRemove(ctx, containerName, types.ContainerRemoveOptions{Force: true})
+		return err
+	}
+
+	if err := r.file.Write(r.nameFile, []byte(containerName)); err != nil {
+		_ = docker.ContainerRemove(ctx, containerName, types.ContainerRemoveOptions{Force: true})
+		return err
+	}
+
+	time.Sleep(15 * time.Second)
+
+	return nil
+}
+
+// StopLibvirt stops a containerized libvirt instance.
+func (r *Runner) StopLibvirt(ctx context.Context) error {
+	name, err := r.file.Read(r.nameFile)
+	if err != nil {
+		if errors.Is(err, afero.ErrFileNotFound) {
+			return nil
+		}
+		return err
+	}
+
+	docker, err := docker.NewClientWithOpts(docker.FromEnv, docker.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+	defer docker.Close()
+	if err := docker.ContainerRemove(ctx, string(name), types.ContainerRemoveOptions{Force: true}); err != nil {
```

Comment:
Does `docker.ContainerRemove()` not return e.g. a NotFoundErr that can be checked, or why do we use a file as proxy if the container has already been deleted/does not exist?

Reply:
The file is not used as proxy to check if the container is still running, but to ensure we remove the correct container during terminate.
Since the terminate command does not know about the name given to cluster resources during create, we need some additional way to keep this information.
Alternatively we could maybe parse the `terraform.tfvars` file to get the name, not sure if this is the best option though

Reply:
I assume that there's no good way of using terraform itself to handle this container? And simply hard coding the name also has drawbacks. Then my best solution would be to to use something like terraform state show to natively get the name variable.

Reply:
I would also agree with doing all this in Terraform if possible. 

This Docker creation in Go seems kinda weird when the metadata-grpc server is created in Terraform, but libvirt is created in Go code. I would prefer keeping this consistent and not scattered across two places.

Reply:
Does not work with Terraform since we require the libvirt container to start before everything else does.
However, setting up the libvirt provider requires the libvirt connection URI. This needs to be set in the main terraform file, meaning we need to know the connection URI when we start applying the terraform templates.
This is not possible if the libvirt container itself is deployed using terraform

Reply:
Mh, there is no dependency management with Terraform and the Docker provider?
Would [defining a different target](https://groups.google.com/g/terraform-tool/c/wKGyI-C03aA/m/zflLYcQLBQAJ) and calling Terraform first on the libvirt target, then the rest of the deployment work?

Reply:
While theoretically possible, this would require a major rewrite to our terraform module.
And I'm unsure if the benefit of not having a separate go module to start the libvirt container is worth the added complexity of the terraform module and qemu terraform template.

Starting the libvirt container as its own instance also gives us the benefit of using the same terraform deployment for containerized libvirt as well as local libvirt installations

Reply:
The last suggestion I thought about was to simply have 2 seperate Terraform files. Since this would necessitate 2 different Terraform workspaces, this is also a sub-optimal solution. Therefore, I'm fine with keeping this part as is. At least until we come up with a solution with no downsides.